### PR TITLE
Figure out what methods should exist between app and worker

### LINKED ISSUE #108: Figure out what methods should exist between the app and the worker!
Right now `compileModel()` and `saveFile()` are wonky. Addtionally, there probably should be events that fire and update stores on the front-end.

Please ask me, @eddieantonio, to clarify my thoughts on this issue, but I think it should be addressed sooner rather than later! With tests!

### PR DESCRIPTION
**Pull request by xqiu1**
Fixes #108

### CODE REVIEW

**Comment by xqiu1** [comment]
@eddieantonio This is just a version to see if I'm on the right track 😅 Are you available to have a look at this?

File: src/worker/predictive-text-studio-worker-impl.ts [hunk 1]
```diff
@@ -39,4 +50,69 @@ export class PredictiveTextStudioWorkerImpl
...
+
+  onPackageCompileError(error: Error, callback: (err: Error) => void): void {
+    // TODO
+    callback(error);
+  }
+
+  onPackageCompileSuccess(kmp: Blob, callback: (kmp: Blob) => void): void {
+    // TODO
+    callback(kmp);
+  }
```
Comment (xqiu1) [inline src/worker/predictive-text-studio-worker-impl.ts]: I'm not very sure about what should be done within these methods.

File: src/common/predictive-text-studio-worker.ts [hunk 2]
```diff
@@ -15,4 +18,35 @@ export interface PredictiveTextStudioWorker {
...
   updateBCP47Tag(bcp47Tag: string): Promise<void>;
+  /**
+   * Generate a KMP file and return a blob of it
+   */
+  generateKMPFromStorage(): Promise<Blob | void>;
+  /**
+   * Take a dictionary source and store it
+   * @param name
+   * @param contents
+   * Return how many words were added
```
Comment (eddieantonio) [inline src/common/predictive-text-studio-worker.ts]:
```suggestion
   * @return how many words were added
```

File: src/common/predictive-text-studio-worker.ts [hunk 3]
```diff
@@ -15,4 +18,35 @@ export interface PredictiveTextStudioWorker {
    * @param bcp47Tag
    */
   updateBCP47Tag(bcp47Tag: string): Promise<void>;
+  /**
+   * Generate a KMP file and return a blob of it
+   */
+  generateKMPFromStorage(): Promise<Blob | void>;
```
Comment (eddieantonio) [inline src/common/predictive-text-studio-worker.ts]: Why was this method added? Who will call it and when?

File: src/common/predictive-text-studio-worker.ts [hunk 4]
```diff
@@ -15,4 +18,35 @@ export interface PredictiveTextStudioWorker {
...
+  /**
+   * Register a callback that is called when the KMP package has succesfully compiled
+   * @param callback
+   */
+  onPackageCompileSuccess(kmp: Blob, callback: (kmp: Blob) => void): void;
+  /**
+   * Sets optional and required metadata such as BCP-47, language name, author name, copyright string, etc.
+   * @param metadata
+   */
+  setProjectData(metadata: Readonly<RelevantKmpOptions>): void;
```
Comment (eddieantonio) [inline src/common/predictive-text-studio-worker.ts]: I like this signature! Though I think it's better to add `Partial` as well:

```suggestion
  setProjectData(metadata: Partial<Readonly<RelevantKmpOptions>>): void;
```

File: src/worker/predictive-text-studio-worker-impl.ts [hunk 5]
```diff
@@ -39,4 +50,69 @@ export class PredictiveTextStudioWorkerImpl
   updateBCP47Tag(bcp47Tag: string): Promise<void> {
     return this.storage.updateBCP47Tag(bcp47Tag);
   }
+
+  async generateKMPFromStorage(): Promise<Blob | void> {
```
Comment (eddieantonio) [inline src/worker/predictive-text-studio-worker-impl.ts]: I like this method, but I think it should be `private` and **not** part of the App/Worker interface. 
```suggestion
  private async generateKMPFromStorage(): Promise<Blob | void> {
```

File: src/worker/predictive-text-studio-worker-impl.ts (hunk 1 above)
Comment (eddieantonio) [reply src/worker/predictive-text-studio-worker-impl.ts]: They should *register a callback* — aka, store the callback as a variable/property. These callbacks are called later, e.g., in `generateKMPFromStorage()`

```suggestion
  private _emitPackageCompileStart: () => void = () => {};
  private _emitPackageCompileError: (err: Error) => void = () => {};
  private _emitPackageCompileSuccess: (kmp: Blob) => void = () => {};

  onPackageCompileStart(callback: () => void): void {
    this._emitPackageCompileStart = callback;
  }

  onPackageCompileError(callback: (err: Error) => void): void {
    this._emitPackageCompileError = callback;
  }

  onPackageCompileSuccess(callback: (kmp: Blob) => void): void {
    this._emitPackageCompileSuccess = callback;
  }
```

File: src/worker/predictive-text-studio-worker-impl.ts [hunk 6]
```diff
@@ -39,4 +50,69 @@ export class PredictiveTextStudioWorkerImpl
...
+    // TODO: Parse multiple dictionary sources, right now just reading the first file
+    this.onPackageCompileStart((): void => {
+      console.log("Compile starts.");
+    });
```
Comment (eddieantonio) [inline src/worker/predictive-text-studio-worker-impl.ts]: Call the registered callback instead:

```suggestion
    this._emitPackageCompileStart();
```

File: src/common/predictive-text-studio-worker.ts [hunk 7]
```diff
@@ -15,4 +18,35 @@ export interface PredictiveTextStudioWorker {
...
+  /**
+   * Register a callbcak that is called if the KMP package compilation fails
+   * @param callback
+   */
+  onPackageCompileError(error: Error, callback: (err: Error) => void): void;
+  /**
+   * Register a callback that is called when the KMP package has succesfully compiled
+   * @param callback
+   */
+  onPackageCompileSuccess(kmp: Blob, callback: (kmp: Blob) => void): void;
```
Comment (eddieantonio) [inline src/common/predictive-text-studio-worker.ts]:
```suggestion
  onPackageCompileError(callback: (err: Error) => void): void;
  /**
   * Register a callback that is called when the KMP package has succesfully compiled
   * @param callback
   */
  onPackageCompileSuccess(callback: (kmp: Blob) => void): void;
```

File: src/worker/predictive-text-studio-worker-impl.ts [hunk 8]
```diff
@@ -39,4 +50,69 @@ export class PredictiveTextStudioWorkerImpl
...
+
+    const storedFiles = await this.storage.fetchAllFiles();
+    if (storedFiles.length < 1) {
+      const error = new Error("Cannot find any file in the IndexedDB");
+      this.onPackageCompileError(error, (error: Error): void => {
+        console.log(error);
+      });
```
Comment (eddieantonio) [inline src/worker/predictive-text-studio-worker-impl.ts]:
```suggestion
      this._emitPackageCompileError(new Error("Cannot find any files in the IndexedDB"));
```

File: src/worker/predictive-text-studio-worker-impl.ts [hunk 9]
```diff
@@ -39,4 +50,69 @@ export class PredictiveTextStudioWorkerImpl
...
+    } else {
+      const kmpFile = await linkStorageToKmp(this.storage);
```
Comment (eddieantonio) [inline src/worker/predictive-text-studio-worker-impl.ts:69]: `linkStorageToKMP()` can raise an error which _should_ be caught and passed to the `this. _emitPackageCompileError` callback.

File: src/worker/predictive-text-studio-worker-impl.ts [hunk 10]
```diff
@@ -39,4 +50,69 @@ export class PredictiveTextStudioWorkerImpl
...
+      const blob = new Blob([kmpFile], { type: "application/octet-stream" });
+      this.onPackageCompileSuccess(blob, (kmp: Blob): void => {
+        console.log("Compile successes.");
+      });
+      return blob;
```
Comment (eddieantonio) [inline src/worker/predictive-text-studio-worker-impl.ts]:
```suggestion
      this._emitPackageCompileSuccess(blob);
```

File: src/worker/storage.ts [hunk 11]
```diff
@@ -28,18 +28,51 @@ export interface StoredWordList {
...
+  /**
+   *
+   */
+  version?: string;
+}
 
 export class PredictiveTextStudioDexie extends Dexie {
   files: Dexie.Table<StoredWordList, number>;
   packageInfo: Dexie.Table<StoredPackageInfo, number>;
+  projectData: Dexie.Table<StoredProjectData, number>;
```
Comment (eddieantonio) [inline src/worker/storage.ts:64]: Why did you create a new table instead of extending the existing table?

**Review by eddieantonio — CHANGES_REQUESTED** [review]

File: src/app/components/BCP47Tag.svelte [hunk 12]
```diff
@@ -12,7 +12,7 @@
   $: displayEnterBCP47Tag = bcp47Tag == "" || validTag;
   $: {
     if (validTag) {
-      worker.updateBCP47Tag(bcp47Tag);
+      worker.setProjectData({ languages: [{ name: "", id: bcp47Tag }] });
```
Comment (xqiu1) [inline src/app/components/BCP47Tag.svelte:15]: We may use the bcp47Tag to generate the language name, or in an opposite way (let the user to input language name and we generate the bcp47Tag).

**Review by IanSeng — APPROVED** [review]

**Review by eddieantonio — APPROVED** [review]

**Pull request merged**